In [3]:
import pandas as pd
import numpy as np
import os

#02 Import data

In [5]:
# Importing the "products.csv" data set
df_prods = pd.read_csv(r'/Users/peterguan/Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [6]:
path = R'/Users/peterguan/Instacart Basket Analysis'

In [7]:
path

'/Users/peterguan/Instacart Basket Analysis'

In [116]:
# Importing the "orders.csv" data set
df_ords = pd.read_csv(r'/Users/peterguan/Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [9]:
# Listing the descriptive statistics for the 'df_ords' dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#03 Creating a new dataframe

In [11]:
# Creating a new dataframe to test
df_test = pd.DataFrame()

In [12]:
# Creates a new column named 'mix' within 'df_test' and fills it with numeric, string, and boolean values
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
# Checking
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# Checking whether a dataframe contains (mix-typed columns)(code structure: for-loop)
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/j8/l8hsy2p15dj9phf2ztxh0b040000gn/T/ipykernel_29705/2764766494.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [15]:
# Changing the 'mix' column's datatype to a 'str'
df_test['mix'] = df_test['mix'].astype('str')

In [16]:
# Assigns the 'isnull()' function to 'df_prods' dataframe and find missing observations, with “observations” here referring to entries in your dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# Create a subset of the dataframe by creating a new variable named 'df_nan' that lists all the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

# Replacing Missing Values with mean or median
df['column with missings'].fillna(mean value, inplace=True)
df['column with missings'].fillna(median value, inplace=True)

In [21]:
# Checking the number of rows and columns
df_prods.shape

(49693, 5)

In [22]:
# Creating a new dataframe that shows non-missing values 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
# Checking if the 16 row was omitted
df_prods_clean.shape

(49677, 5)

In [24]:
# Dropping all missing values AND also OVERWRITING 'df_prods' with a new version due to 'inplace = True'
df_prods.dropna(inplace = True)

In [25]:
df_prods.shape

(49677, 5)

In [26]:
# Dropping all missing values from a particular column AND also OVERWRITING
df_prods.dropna(subset = ['product_name'], inplace = True)

In [27]:
df_prods.shape

(49677, 5)

In [28]:
# Looks for FULL DUPLICATES within the 'df_prods_clean' dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
# Deleting duplicates from the 'df_prods_clean' dataframe and creating a new variable
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [31]:
df_prods_clean_no_dups.shape

(49672, 5)

In [32]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [33]:
# Checking WHY the 'max' price is 99999
df_max_price = df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 1000]

In [34]:
df_max_price

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [35]:
# Replacing the 99999 price with 4 by researching the price of the product on Google
df_prods_clean_no_dups = df_prods_clean_no_dups.replace(99999, 4)

In [36]:
# Replacing the 14900 price with 3 by researching the price of the product on Google
df_prods_clean_no_dups = df_prods_clean_no_dups.replace(14900, 3)

In [37]:
# Rechecking if the 'MAX' price is correct now
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.049867     67.762442      11.728942      7.680269
std    14341.069149     38.315784       5.850779      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12431.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [122]:
df_prods_no_delete = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'].isnull() == False]

#04 Exporting dataframe


In [124]:
df_prods_no_delete

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [39]:
# Exporting the df_prods_clean_no_dups dataframe to the 'Prepared data' folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

In [41]:
# Question 2: Running descriptive statistics on the 'df_ords' dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

I do not see anything that looks off or should be investigated further in this data output.

In [43]:
# Question #3: Checking for mixed-type data in the 'df_ords' dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [44]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [45]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

There are no mixed-type data in the 'df_ords' dataframe

In [47]:
# Question 4: There are no mixed-type data


In [48]:
# Question 5: Checking for missing values in the 'df_ords' dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in the 'days_since_prior_order' column. An explanation for this is: missing values can occur for two reasons: 1) data corruption, or 2) they were never recorded in the first place.

In [50]:
# Question 6: Addressing the missing values by dropping all missing values
# Created a subset of the dataframe by creating a new variable named 'df_missing_values' that lists all the missing values 
df_missing_values = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [51]:
df_missing_values

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

# The column 'days_since_prior_order' has a MISSING VALUE because it is these user's first ever order as shown by their
'order_number'. When you check each 'user_id' with the following code: df_check_customer = df_missing_values.loc[df_missing_values['user_id'] == 133]


In [53]:
# Checking the rows x columns of the 'df_ords' dataframe WITH the MISSING VALUES
df_ords.shape

(3421083, 7)

In [54]:
# Creating a new dataframe that shows non-missing values
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [55]:
# Checking the rows x columns of the 'df_ords' dataframe WITHOUT the MISSING VALUES
df_ords_clean.shape

(3214874, 7)

In [56]:
# Question 7: Checking for duplicates
df_duplicates = df_ords_clean[df_ords_clean.duplicated]

In [57]:
df_duplicates

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are NO duplicates in the 'df_ords_clean' dataframe.

In [59]:
# Question 8: There are NO duplicates in the 'df_ords_clean' dataframe.

In [60]:
# Question 9: Exporting the df_ords_clean dataframe to the 'Prepared data' folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))


In [61]:
df_check_customer = df_missing_values.loc[df_missing_values['user_id'] == 133]

In [62]:
df_check_customer

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
1915        1915   3187763      133             1                   4   

      order_hour_of_day  days_since_prior_order  
1915                 19                     NaN